client/manager/manager.go

```diff
@@ -0,0 +1,217 @@
+package manager
+
+import (
+	"encoding/base64"
+	"fmt"
+
+	"errors"
+
+	"github.com/coreos/dex/client"
+	pcrypto "github.com/coreos/dex/pkg/crypto"
+	"github.com/coreos/dex/pkg/log"
+	"github.com/coreos/dex/repo"
+	"github.com/coreos/go-oidc/oidc"
+	"golang.org/x/crypto/bcrypt"
+)
+
+const (
+	// Blowfish, the algorithm underlying bcrypt, has a maximum
+	// password length of 72. We explicitly track and check this
+	// since the bcrypt library will silently ignore portions of
+	// a password past the first 72 characters.
+	maxSecretLength = 72
+)
+
+type SecretGenerator func() ([]byte, error)
+
+func DefaultSecretGenerator() ([]byte, error) {
+	return pcrypto.RandBytes(maxSecretLength)
+}
+
+func CompareHashAndPassword(hashedPassword, password []byte) error {
+	if len(password) > maxSecretLength {
+		return errors.New("password length greater than max secret length")
+	}
+	return bcrypt.CompareHashAndPassword(hashedPassword, password)
+}
+
+// ClientManager performs client-related "business-logic" functions on client and related objects.
+// This is in contrast to the Repos which perform little more than CRUD operations.
+type ClientManager struct {
+	clientRepo        client.ClientRepo
+	begin             repo.TransactionFactory
+	secretGenerator   SecretGenerator
+	clientIDGenerator func(string) (string, error)
+}
+
+type ManagerOptions struct {
+	SecretGenerator   func() ([]byte, error)
+	ClientIDGenerator func(string) (string, error)
+}
+
+func NewClientManager(clientRepo client.ClientRepo, txnFactory repo.TransactionFactory, options ManagerOptions) *ClientManager {
+	if options.SecretGenerator == nil {
+		options.SecretGenerator = DefaultSecretGenerator
+	}
+	if options.ClientIDGenerator == nil {
+		options.ClientIDGenerator = oidc.GenClientID
+	}
+	return &ClientManager{
+		clientRepo:        clientRepo,
+		begin:             txnFactory,
+		secretGenerator:   options.SecretGenerator,
+		clientIDGenerator: options.ClientIDGenerator,
+	}
+}
+
+func NewClientManagerFromClients(clientRepo client.ClientRepo, txnFactory repo.TransactionFactory, clients []client.Client, options ManagerOptions) (*ClientManager, error) {
+	clientManager := NewClientManager(clientRepo, txnFactory, options)
+	tx, err := clientManager.begin()
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	for _, c := range clients {
+		if c.Credentials.Secret == "" {
+			return nil, fmt.Errorf("client %q has no secret", c.Credentials.ID)
+		}
+
+		cli, err := clientManager.clientFromMetadata(c.Metadata)
+		if err != nil {
+			return nil, err
+		}
+		cli.Admin = c.Admin
+
+		_, err = clientRepo.New(tx, cli)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+	return clientManager, nil
+}
+
+func (m *ClientManager) New(meta oidc.ClientMetadata) (*oidc.ClientCredentials, error) {
+	tx, err := m.begin()
```

Comment:
Because you have New() taking a `ClientMetadata` you can't set DexAdmin and other non-metadata properites on it in the same transaction